mithril-common/src/chain_reader/pallas_chain_reader.rs

```diff
@@ -0,0 +1,262 @@
+            NextResponse::RollForward(raw_block, forward_tip) => {
+                Ok(Some(ChainBlockNextAction::RollForward {
+                    next_point: forward_tip.into(),
+                    raw_block,
+                }))
+            }
+            NextResponse::RollBackward(rollback_point, _) => {
+                Ok(Some(ChainBlockNextAction::RollBackward {
+                    rollback_point: rollback_point.into(),
+                }))
+            }
+            NextResponse::Await => Ok(None),
+        }
+    }
+
+    async fn get_next_chain_block(
+        &mut self,
+        point: &ChainPoint,
+    ) -> StdResult<Option<ChainBlockNextAction>> {
+        let client = self.get_client().await?;
+        let chainsync = client.chainsync();
+
+        chainsync
+            .find_intersect(vec![point.to_owned().into()])
+            .await?;
+
+        let next = match chainsync.has_agency() {
+            true => chainsync.request_next().await?,
+            false => chainsync.recv_while_must_reply().await?,
+        };
+
+        self.process_next_chain_block(next).await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::fs;
+
+    use pallas_network::{
+        facades::NodeServer,
+        miniprotocols::{
+            chainsync::{BlockContent, Tip},
+            Point,
+        },
+    };
+    use tokio::net::UnixListener;
+
+    use super::*;
+
+    use crate::test_utils::TempDir;
+
+    /// Returns a fake chain point for testing purposes.
+    fn get_fake_chain_point_backwards() -> ChainPoint {
+        ChainPoint::from(Point::Specific(
+            1654413,
+            hex::decode("7de1f036df5a133ce68a82877d14354d0ba6de7625ab918e75f3e2ecb29771c2")
+                .unwrap(),
+        ))
+    }
+
+    /// Returns a fake chain point for testing purposes.
+    fn get_fake_chain_point_forwards() -> ChainPoint {
+        Tip(
+            Point::Specific(
+                1654413,
+                hex::decode("7de1f036df5a133ce68a82877d14354d0ba6de7625ab918e75f3e2ecb29771c2")
+                    .unwrap(),
+            ),
+            1337,
+        )
+        .into()
+    }
+
+    /// Returns a fake intersection point for testing purposes.
+    fn get_fake_intersection_point() -> Point {
+        Point::Specific(
+            1654413,
+            hex::decode("7de1f036df5a133ce68a82877d14354d0ba6de7625ab918e75f3e2ecb29771c2")
+                .unwrap(),
+        )
+    }
+
+    /// Creates a new work directory in the system's temporary folder.
+    fn create_temp_dir(folder_name: &str) -> PathBuf {
+        TempDir::create_with_short_path("pallas_chain_observer_test", folder_name)
+    }
+
+    /// Sets up a mock server for related tests.
+    ///
+    /// Use the `intersections` parameter to define exactly how many
+    /// local state queries should be intersepted by the `mock_server`
+    /// and avoid any panic errors.
+    async fn setup_server(socket_path: PathBuf, action: String) -> tokio::task::JoinHandle<()> {
```

Comment:
IMO the test server setup should be given list of `ChainBlockNextAction` and should serve these actions to the client. This would avoid using an action and would help testing more sophisticated scenarios.

Also, the action should probably be an an enum instead of a string.